### PR TITLE
Add a changelog for x.py and nag contributors until they read it

### LINKED ISSUE #76617: Add a changelog to `x.py`
This would make it easier to find and learn more about breaking changes (or even just new features). It would also give us something to point to after a change, especially for changes not large enough to need their own blog post on inside-rust.

I'm imagining it would go in `src/bootstrap/CHANGELOG.md`, following https://keepachangelog.com/.

### PR DESCRIPTION
**Pull request by jyn514**
Add a changelog for x.py

- Add a changelog and instructions for updating it
- Use `changelog-seen` in `config.toml` and `VERSION` in bootstrap to determine whether the changelog has been read.  There's no way to tie reading the changelog to updating the version, so unfortunately they still have to update `config.toml` manually. Actually reading the changelog is optional, anyone can set `changelog-seen = N` without reading (although it's not recommended).
- Nag people if they haven't read the x.py changelog
  + Print message twice to make sure it's seen
- Give different error messages depending on whether the version needs to be updated or added

Closes https://github.com/rust-lang/rust/issues/76617
r? @Mark-Simulacrum

### CODE REVIEW

**Comment by jyn514** [comment]
Example output:
```
$ x.py check
Updating only changed submodules
Submodules updated in 0.02 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
warning: x.py has made several changes in the past month you may want to look at
help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`
note: to silence this warning, add `version = 0` to `config.toml`
Checking std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Checking core v0.0.0 (/home/joshua/rustc/library/core)
^C  Building [==============>                                           ] 9/33: ...
```

File: config.toml.example
Comment (pickfire) [inline config.toml.example:11]: Should we add comment for it?

Comment (jyn514) [reply config.toml.example:11]: Yes, good idea. I'll add that.

File: src/bootstrap/bin/main.rs
Comment (pickfire) [inline src/bootstrap/bin/main.rs:34]: Where should the user look at?

Comment (jyn514) [reply src/bootstrap/bin/main.rs:34]: It prints that a few lines later; see https://github.com/rust-lang/rust/pull/76626#issuecomment-691392248 for an example.

File: config.toml.example
Comment (pickfire) [inline config.toml.example:12]:
```suggestion
# Keeps track of the latest version of `x.py` used.
```

File: config.toml.example
Comment (pickfire) [inline config.toml.example:13]:
```suggestion
# If it does not match the x.py that is currently running,
```
We should not use `doesn't` since this is written text.

File: src/bootstrap/CHANGELOG.md
Comment (pickfire) [inline src/bootstrap/CHANGELOG.md:6]: I believe we should have a section here for unreleased changes?

File: src/bootstrap/bin/main.rs
Comment (pickfire) [reply src/bootstrap/bin/main.rs:34]: Yes, but why does it needs this lines? The few lines below already explained it well.

File: src/bootstrap/CHANGELOG.md
Comment (jyn514) [reply src/bootstrap/CHANGELOG.md:6]: There is no such thing as unreleased changes, because there are no releases. Part of the point of nagging you to read the changelog is that there's no way to opt-out of breaking changes.

File: src/bootstrap/bin/main.rs
Comment (jyn514) [reply src/bootstrap/bin/main.rs:34]: I don't think they do?
```
help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`
note: to silence this warning, add `version = 0` to `config.toml`
```
That doesn't say _what_ changed - this line specifically says that _x.py_ changed.

File: config.toml.example
Comment (jyn514) [reply config.toml.example:12]: I don't think this is a good description .. the changes would happen with or without this line, this just makes you aware of them.

Comment (pickfire) [reply config.toml.example:12]: Okay, maybe I don't put the reason.
```suggestion
# Keeps track of the latest version of `x.py` used.
```
Removes redundant `This` (it couldn't be `that` right?) and `you've` (`you` may not be good and there is only one person here).

File: config.toml.example
Comment (pickfire) [inline config.toml.example:14]:
```suggestion
# x.py will warn if it does not match the version of `x.py` running.
```

File: src/bootstrap/CHANGELOG.md
Comment (pickfire) [reply src/bootstrap/CHANGELOG.md:6]: > Changes that do not affect contributors to the compiler or users
> building rustc from source don't need an update to `VERSION` (although they
> should still update the `CHANGELOG`).

Oh, I meant where to update the `CHANGELOG` if there is no breaking changes?

File: src/bootstrap/bin/main.rs
Comment (pickfire) [reply src/bootstrap/bin/main.rs:34]:
```
warning: there have been changes to x.py since you last updated `version = 0`
help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`
note: to silence this warning, add `version = 1` to `config.toml`
```
Wouldn't user get the same warning if they already have `version = 1` in config.toml? I believe it should show `add version = 0 to` only if version does not exists but `change version = 0 in` if config already exists.

File: src/bootstrap/bin/main.rs
Comment (pickfire) [inline src/bootstrap/bin/main.rs:41]:
```suggestion
    if let Some(version) = config.version {
        if version != VERSION {
            println!("warning: x.py have new version but {} is used", version);
            println!("help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`");
            println!("note: to silence this warning, use `version = {}` in `config.toml`", VERSION);
        }
    } else {
        println!("warning: x.py have new changes you may want to look at");
        println!("help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`");
        println!("note: to silence this warning, add `version = {}` to `config.toml`", VERSION);
    }
```
Suggestion than I mentioned above.

File: src/bootstrap/bin/main.rs
Comment (jyn514) [reply src/bootstrap/bin/main.rs:34]: > Wouldn't user get the same warning if they already have version = 1 in config.toml?

No, that's what the `version != VERSION` check is doing. I could probably improve the error message though.

Comment (jyn514) [reply src/bootstrap/bin/main.rs:34]: Ok, now it will say 'update `config.toml` to use `version = 1`' instead.

File: src/bootstrap/bin/main.rs
Comment (pickfire) [inline src/bootstrap/bin/main.rs:22]: Why should we allocate a `String` here when we don't need to? We could just duplicate the print here right?

File: src/bootstrap/CHANGELOG.md
Comment (jyn514) [reply src/bootstrap/CHANGELOG.md:6]: Ah good point. Maybe they should just be added to the old version? That breaks the assumption that versions are immutable ... I'll add an `unreleased` section like you mentioned originally.

File: src/bootstrap/bin/main.rs
Comment (jyn514) [reply src/bootstrap/bin/main.rs:22]: This is a very tiny fraction of the runtime ... just printing it will be 100x as slow as the allocation.

File: src/bootstrap/bin/main.rs
Comment (jyn514) [reply src/bootstrap/bin/main.rs:34]: Example output:
```
warning: there have been changes to x.py since you last updated `version = 1`
help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`
note: to silence this warning, update `config.toml` to use `version = 0` instead
```

File: src/bootstrap/config.rs
Comment (pickfire) [inline src/bootstrap/config.rs:524]: This is appearing right after toml which I think should be separated.
```suggestion

        config.version = toml.version;
```

**Review by pickfire — APPROVED** [review]
Looks good to me except the last thing mentioned.

File: src/bootstrap/README.md
Comment (Mark-Simulacrum) [inline src/bootstrap/README.md:346]: Hm, what would an example of such a change be? It feels like if it affects no one, it should not be in the changelog...

File: src/bootstrap/bin/main.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/bin/main.rs:34]: "month" here feels a bit odd -- maybe "since the last version"? Or "recently"?

File: src/bootstrap/bin/main.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/bin/main.rs:25]:
```suggestion
                "warning: there have been changes to x.py since you last updated.",
```

Comment (Mark-Simulacrum) [reply src/bootstrap/bin/main.rs:25]: I don't think it's all that helpful to tell them the current version in config.toml, or at least, I wasn't able to find a nice way to fit it in here without being very verbose or misleading. The current PR seems too "brief" -- it looks weird to just have a version = {} and is unclear to me which version that is.

File: src/bootstrap/config.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/config.rs:264]: We should require this, I think? Is there a reason not to?

**Review by Mark-Simulacrum — COMMENTED** [review]
We should make sure that configure generates the appropriate version such that you don't immediately get nagged.

File: config.toml.example
Comment (jyn514) [reply config.toml.example:16]: That seems a little long, maybe `changelog-seen`? I agree `version` is a little misleading.

File: src/bootstrap/README.md
Comment (jyn514) [reply src/bootstrap/README.md:346]: Good point, I'll remove that bit.

Do you still think this should warn when a new configuration option is added? That's not a breaking change, although we still might want to tell people about it.

File: src/bootstrap/bin/main.rs
Comment (jyn514) [reply src/bootstrap/bin/main.rs:25]: That makes sense, we're telling them to update `version = VERSION` a few lines later anyway so they know the version and this message are related.

File: src/bootstrap/config.rs
Comment (jyn514) [reply src/bootstrap/config.rs:264]: If it's mandatory, that breaks everyone with `config.toml` currently (a hard break, the TOML won't parse). If we make it optional, they just get a warning to look at the changelog.

**Review by pickfire — APPROVED** [review]
Looks good to me.

**Comment by jyn514** [comment]
> We should make sure that configure generates the appropriate version such that you don't immediately get nagged.

Missed this comment the first time, sorry. I'm not sure whether that's helpful - distro maintainers (cc @infinity0) will also want to know if x.py had major changes. If `configure` silently sets `changelog-seen` to the latest version, they won't see the difference.

If they _have_ seen the changes, they can always pass `--set version=1`.

**Comment by Mark-Simulacrum** [comment]
Hm, I guess that's an interesting point. Probably given that we regularly generate hundreds of log lines anyway a warning at the beginning doesn't matter too much...

**Comment by jyn514** [comment]
We could add _another_ option to make the warning fatal? 😆

**Comment by jyn514** [comment]
> given that we regularly generate hundreds of log lines anyway

Might be useful to look into reducing that ... but that's separate from this change.

**Comment by jyn514** [comment]
ping @Mark-Simulacrum - is this waiting on anything from me? Did you want me to implement the (optional) hard error if the versions don't match?

File: src/bootstrap/README.md
Comment (Mark-Simulacrum) [inline src/bootstrap/README.md:319]:
```suggestion
* Update `changelog-seen = N` in `config.toml.example`.
```

File: src/bootstrap/bin/main.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/bin/main.rs:15]: Can we re-print this at the end of the build? I worry that in the not atypical case of quick screen-filling output the helpful message will otherwise get lost. And in cases where very little output is given, it seems not horrible to have a couple extra lines, maybe with some kind of prefix to make it clear it's a repeat.

**Comment by Mark-Simulacrum** [comment]
Okay, I would appreciate:
 * Squashing the commits down
 * Updating the PR description, looks like some stale notes of version still

Also left a few nits/suggestions.

File: src/bootstrap/bin/main.rs
Comment (jyn514) [reply src/bootstrap/bin/main.rs:15]: I changed this to print twice, but because rustbuild uses `std::process::exit`, it won't get printed if an error occurs.

Successful message:
```
$ x.py test tidy
warning: x.py has made several changes recently you may want to look at
help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`
note: to silence this warning, add `changelog-seen = 1` to `config.toml`
Building stage0 tool tidy (x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.08s
tidy check
* 617 error codes
* highest error code: E0774
Checking which error codes lack tests...
Found 430 error codes
Found 0 error codes with no tests
Done!
* 298 features
fmt check
skip untracked path src/bootstrap/defaults/ during rustfmt invocations
skip untracked path src/tmp.txt during rustfmt invocations
skip untracked path tmp.rs during rustfmt invocations
skip untracked path x.rs during rustfmt invocations
skip untracked path y.rs during rustfmt invocations
warning: x.py has made several changes recently you may want to look at
help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`
note: to silence this warning, add `changelog-seen = 1` to `config.toml`
note: this message was printed twice to make it more likely to be seen
Build completed successfully in 0:00:06
```

Message on error:
```
warning: x.py has made several changes recently you may want to look at
help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`
note: to silence this warning, add `changelog-seen = 1` to `config.toml`
Building stage0 tool tidy (x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.09s
tidy check
* 617 error codes
* highest error code: E0774
Checking which error codes lack tests...
Found 430 error codes
Found 0 error codes with no tests
Done!
tidy error: /home/joshua/rustc/src/bootstrap/bin/main.rs:6: trailing whitespace
some tidy checks failed


command did not execute successfully: "/home/joshua/rustc/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/home/joshua/rustc" "/home/joshua/rustc/build/x86_64-unknown-linux-gnu/stage0/bin/cargo"
expected success, got: exit code: 1


failed to run: /home/joshua/rustc/build/bootstrap/debug/bootstrap test tidy
Build completed unsuccessfully in 0:00:03
```

Maybe it could be moved to `bootstrap.py` or `x.py` instead?

Comment (Mark-Simulacrum) [reply src/bootstrap/bin/main.rs]: Hm let's avoid that for now.

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/76987/checks?check_run_id=1141645171)

**Comment by jyn514** [comment]
@Dylan-DPC it's a merge conflict with https://github.com/rust-lang/rust/issues/76165, I can't fix it until that PR's merged.

**Comment by jyn514** [comment]
@bors r=Mark-Simulacrum

Forgot to add https://github.com/rust-lang/rust/pull/76628 to the changelog.

**Comment by infinity0** [comment]
@jyn514 I just noticed this, sorry I didn't find time to reply to your other messages. Thanks for pinging, this will be fine. Moving forward the changelog ought to omit minor things (that don't affect external behaviour) otherwise it would be redundant wrt `git log`.

**Pull request merged**
